File: src/main/resources/META-INF/resources/frontend/idle-notification.js
/*
 * Copyright 2000-2020 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { html, PolymerElement } from '@polymer/polymer/polymer-element.js';
import { ThemableMixin } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
import '@vaadin/vaadin-dialog/src/vaadin-dialog.js';
import '@vaadin/vaadin-button/src/vaadin-button.js';
import './idle-notification-shadow-styles.js';

class IdleNotification extends ThemableMixin(PolymerElement) {
  static get properties() {
    return {
      /**
       * Set to true when the notification is open
       * @type {boolean}
       */
      opened: {
        type: Boolean,
        value: false,
        notify: true,
        reflectToAttribute: true,
        observer: '_openedChanged',
      },

      /**
       * Determines whether a close button is displayed to the user
       * @type {boolean}
       */
      closeButtonEnabled: {
        type: Boolean,
        value: false,
        notify: true,
        observer: '_closeButtonEnabledChanged',
      },

      /**
       * Determines whether an extend-session button is displayed to the user
       * @type {boolean}
       */
      extendSessionButtonEnabled: {
        type: Boolean,
        value: false,
        notify: true,
        observer: '_extendSessionButtonEnabledChanged',
      },

      /**
       * Determines whether a re-direct button is displayed to the user
       * @type {boolean}
       */
      redirectButtonEnabled: {
        type: Boolean,
        value: false,
        notify: true,
        observer: '_redirectButtonEnabledChanged',
      },

      /**
       * The message displayed to the user before the session expires
       * @type {string}
       */
      beforeExpiredMessage: {
        type: String,
        value: '',
        notify: true,
        observer: '_beforeExpiredMessageChanged',
      },

      /**
       * The caption of the extend-session button
       * @type {string}
       */
      extendSessionButtonCaption: {
        type: String,
        value: 'Stay logged In',
      },

      /**
       * The caption of the re-direct button
       * @type {string}
       */
      redirectButtonCaption: {
        type: String,
        observer: '_redirectButtonCaptionChanged',
      },

      /**
       * The url to redirect the page to upon re-direct-button click
       * @type {string}
       */
      redirectButtonUrl: {
        type: String,
        observer: '_redirectButtonUrlChanged',
      },

      /**
       * The url to redirect the page to upon session timeout
       * @type {string}
       */
      redirectAtTimeoutUrl: {
        type: String,
      },

      /**
       * Determines whether the user is redirected to the redirectUrl upon session timeout
       *
       * NOTE: this would only have an effect if redirectAtTimeoutUrl is set
       * @type {boolean}
       */
      redirectAtTimeoutEnabled: {
        type: Boolean,
        value: true,
      },

      /**
       * The message displayed to the user after the session expires
       * @type {string}
       */
      afterExpiredMessage: {
        type: String,
        value: '',
      },

      /**
       * The number of seconds until the session timesout
       * @type {number}
       */
      maxInactiveInterval: {
        type: Number,
        observer: '_maxInactiveIntervalChanged',
      },

      /**
       * The number of seconds before session timesout at which point
       * the notification will be displayed to the user
       * @type {number}
       */
      secondsBeforeNotification: {
        type: Number,
        observer: '_secondsBeforeNotificationChanged',
      },

      /**
       * Set to true if the time-settings are valid
       * @type {boolean}
       */
      activated: {
        type: Boolean,
        value: false,
        notify: true,
        reflectToAttribute: true,
        observer: '_activatedChanged',
      },

      /**
       * Determines whether an outside click would poke the server (thereby extending the session)
       * @type {boolean}
       */
      extendSessionOnOutsideClick: {
        type: Boolean,
        value: true,
      },

      /**
       * Determines whether an outside click would close the notification.
       *
       * NOTE: this would only have an effect if (extendSessionOnOutsideClick == false)
       *
       * @type {boolean}
       */
      closeNotificationOnOutsideClick: {
        type: Boolean,
        value: false,
      },

      /** @private */
      _timeoutObj: Object,

      /** @private */
      _messageContent: String,

      /** @private */
      _actionsVisibility: {
        type: String,
        value: 'hidden',
      },

      /** @private */
      _redirectButtonVisibility: {
        type: String,
        value: 'hidden',
      },

      /** @private */
      _extendSessionButtonVisibility: {
        type: String,
        value: 'hidden',
      },

      /** @private */
      _headerVisibility: {
        type: String,
        value: 'hidden',
      },

      /** @private */
      _displayProcessStarted: Boolean,

      /** @private */
      _dialogElement: Object,

      /** @private */
      _handleLoadListener: {
         type: Object,
         value: null,
      }
    };
  }

  static get template() {
    return html`
      <style>
        [part='container'] {
          --notification-background-color: var(--idle-notification-background-color, #ffc13f);
          --notification-color: var(--idle-notification-color, var(--lumo-contrast));
          color: var(--notification-color);
          display: flex;
          flex-direction: column;
          background-color: var(--notification-background-color);
          padding: 1em;
          border-radius: var(--lumo-border-radius-m);
        }

        [part='message-container'] {
          display: flex;
        }

        #warning-icon {
          display: flex;
          flex-shrink: 0;
          align-self: center;
          margin-right: 1em;
        }

        [part='header'].hidden {
          display: none;
        }

        [part='header'].visible {
          display: flex;
          justify-content: flex-end;
          margin-bottom: 1em;
        }

        #close-icon {
          display: flex;
          font-size: var(--lumo-font-size-xs);
          height: 1em;
          width: 1em;
        }

        [part='message'] {
          display: flex;
        }

        [part='actions'].visible {
          display: flex;
          justify-content: flex-end;
          margin-top: 1em;
        }

        [part='actions'].hidden {
          display: none;
        }

        vaadin-button {
          --fallback-border-radius: calc(var(--lumo-size-m) / 2);
          color: var(--idle-notification-button-color, --notification-color);
          background-color: var(--idle-notification-button-background-color, white);
          border-radius: var(--lumo-border-radius-m, var(--fallback-border-radius));
        }

        #extend-session.visible {
          margin-left: 1em;
        }

        #extend-session.hidden {
          display: none;
        }

        #redirect.hidden {
          display: none;
        }
      </style>

      <vaadin-dialog id="notification-dialog" part="notification-dialog" theme="notification-dialog-theme">
        <template>
          <div part="container" class="idle-notification-container">
            <div part="header" class$="[[_headerVisibility]]">
              <iron-icon icon="vaadin:close-small" id="close-icon" on-click="_handleCloseIconClick"></iron-icon>
            </div>
            <div part="message-container">
              <iron-icon icon="vaadin:exclamation-circle-o" id="warning-icon"></iron-icon>
              <div part="message">[[_messageContent]]</div>
            </div>
            <div part="actions" class$="[[_actionsVisibility]]">
              <vaadin-button id="redirect" on-click="_handleRedirectButtonClick" class$="[[_redirectButtonVisibility]]">
                [[redirectButtonCaption]]
              </vaadin-button>
              <vaadin-button
                id="extend-session"
                on-click="_handleExtendSessionButtonClick"
                class$="[[_extendSessionButtonVisibility]]"
              >
                [[extendSessionButtonCaption]]
              </vaadin-button>
            </div>
          </div>
        </template>
      </vaadin-dialog>
    `;
  }

  /** @protected */
  ready() {
    super.ready();

    this._dialogElement = this.shadowRoot.querySelector('vaadin-dialog');
    this._dialogElement.$.overlay.addEventListener('vaadin-overlay-outside-click', this._handleOutsideClick.bind(this));
    this._dialogElement.noCloseOnOutsideClick = !this.closeNotificationOnOutsideClick;
    this._dialogElement.noCloseOnEsc = true;
    this._dialogElement.modeless = false;

    this._registerXhrListener();
  }

  /** @protected */
  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('load', this._handleLoadListener);
    this._clearTimeoutObject();
    // to not trigger _handleLoad after disconnecting the component
    this._displayProcessStarted = true;
  }

  /** @private */
  _registerXhrListener() {
    let currRequest;
    let origOpen = XMLHttpRequest.prototype.open;
    let thisComponent = this;
    XMLHttpRequest.prototype.open = function () {
      currRequest = this;
      thisComponent._handleLoadListener = (e) => thisComponent._handleLoad(currRequest);
      this.addEventListener('load', thisComponent._handleLoadListener);
      origOpen.apply(this, arguments);
    };
  }

  /** @private */
  _handleLoad(currRequest) {
    if (
        currRequest.status === 200 &&
        !this._displayProcessStarted &&
        this._isVaadinRequest(currRequest)
    ) {
      this._resetTimer();
      if (this.opened) {
        this.opened = false;
      }
    }
  };

  /** @private */
  _closeButtonEnabledChanged(isCloseButtonEnabled, wasCloseButtonEnabled) {
    if (isCloseButtonEnabled) {
      this._headerVisibility = 'visible';
    } else {
      this._headerVisibility = 'hidden';
    }
  }

  /** @private */
  _extendSessionButtonEnabledChanged(isExtendSessionButtonEnabled, wasExtendSessionButtonEnabled) {
    if (isExtendSessionButtonEnabled) {
      this._extendSessionButtonVisibility = 'visible';
      this._actionsVisibility = 'visible';
    } else if (!(this.redirectButtonEnabled && this.redirectButtonCaption && this.redirectUrl)) {
      this._actionsVisibility = 'hidden';
    }
  }

  /** @private */
  _redirectButtonEnabledChanged(isRedirectButtonEnabled, wasRedirectButtonEnabled) {
    this._validateRedirectButton(isRedirectButtonEnabled, this.redirectButtonCaption, this.redirectButtonUrl);
  }

  /** @private */
  _redirectButtonCaptionChanged(newRedirectButtonCaption, oldRedirectButtonCaption) {
    this._validateRedirectButton(this.redirectButtonEnabled, newRedirectButtonCaption, this.redirectButtonUrl);
  }

  /** @private */
  _redirectButtonUrlChanged(newRedirectButtonUrl, oldRedirectButtonUrl) {
    this._validateRedirectButton(this.redirectButtonEnabled, this.redirectButtonCaption, newRedirectButtonUrl);
  }

  /** @private */
  _validateRedirectButton(isRedirectButtonEnabled, redirectButtonCaption, redirectButtonUrl) {
    if (isRedirectButtonEnabled && redirectButtonCaption && redirectButtonUrl) {
      this._redirectButtonVisibility = 'visible';
      this._actionsVisibility = 'visible';
    } else if (!this.extendSessionButtonEnabled) {
      this._actionsVisibility = 'hidden';
    }
  }

  /** @private */
  _activatedChanged(isActivated, wasActivated) {
    if (isActivated && !wasActivated) {
      this._resetTimer();
    } else {
      this._clearTimeoutObject();
    }
  }

  /** @private */
  _handleCloseIconClick(e) {
    this.opened = false;
  }

  /** @private */
  _handleExtendSessionButtonClick(e) {
    this._pokeServer();
    this.dispatchEvent(new CustomEvent('vaadin-idle-notification-extend-session', { bubbles: true, composed: true }));
  }

  /** @private */
  _handleRedirectButtonClick(e) {
    if (this.redirectButtonUrl) {
      this.dispatchEvent(new CustomEvent('vaadin-idle-notification-redirect', { bubbles: true, composed: true }));
      this._doRedirect(this.redirectButtonUrl);
    }
  }

  /** @private */
  _resetTimer() {
    clearTimeout(this._timeoutObj);
    this._timeoutObj = setTimeout(
      (e) => this._displayNotification(e),
      (this.maxInactiveInterval - this.secondsBeforeNotification) * 1000
    );
  }

  /** @private */
  _isVaadinRequest(req) {
    const reqUrl = new URL(req.responseURL);
    // ignore heartbeat requests, so timeout value can be larger than heartbeat interval
    return reqUrl.searchParams.has('v-r') && reqUrl.searchParams.get('v-r') !== 'heartbeat';
  }

  /** @private */
  _displayNotification(e) {
    // The UI may have one of its components (for example, a Textfield) in focus.
    // Opening the dialog may cause such component to lose focus, hence triggering a
    // change event that send a request to the server (and thereby extending the session).
    // The following flag is used to prevent this sequence from happening.
    this._displayProcessStarted = true;

    this.opened = true;
    let timeleft = this.secondsBeforeNotification - 1;
    this._timeoutObj = setInterval((e) => {
      if (timeleft <= 0) {
        clearInterval(this._timeoutObj);
        this._handleSessionTimeout(e);
      } else {
        this._updateFormattedMessage(this.beforeExpiredMessage, timeleft);
        timeleft -= 1;
      }
    }, 1000);
  }

  /** @private */
  _handleSessionTimeout(e) {
    this.dispatchEvent(new CustomEvent('vaadin-idle-notification-timeout', { bubbles: true, composed: true }));

    if (this.redirectAtTimeoutEnabled && this.redirectAtTimeoutUrl) {
      this._doRedirect(this.redirectAtTimeoutUrl);
    } else {
      this._updateFormattedMessage(this.afterExpiredMessage, 0);
      this.extendSessionOnOutsideClick = false;
      this.closeNotificationOnOutsideClick = true;
      this._actionsVisibility = 'hidden';
    }
  }

  /** @private */
  _doRedirect(url) {
    window.location.replace(url);
  }

  /** @private */
  _clearTimeoutObject() {
    if (this._timeoutObj) {
      clearInterval(this._timeoutObj);
    }
  }

  /** @private */
  _openedChanged(opened, wasOpened) {
    if (this._dialogElement) {
      this._dialogElement.opened = opened;
    }
    if (opened) {
      this._updateFormattedMessage(this.beforeExpiredMessage, this.secondsBeforeNotification);
      if(!wasOpened) this.dispatchEvent(new CustomEvent('vaadin-idle-notification-open', { bubbles: true, composed: true }));
    } else {
      this._displayProcessStarted = false;
      if(wasOpened) this.dispatchEvent(new CustomEvent('vaadin-idle-notification-close', { bubbles: true, composed: true }));
    }
  }

  /** @private */
  _beforeExpiredMessageChanged(newValue, oldValue) {
    this._updateFormattedMessage(newValue, this.secondsBeforeNotification);
  }

  /** @private */
  _updateFormattedMessage(message, secondsToTimeout) {
    let SECS_TO_TIMEOUT = secondsToTimeout;
    let SECS_MAX_IDLE_TIMEOUT = this.maxInactiveInterval;
    let minutesToTimeout = Math.floor(SECS_TO_TIMEOUT / 60);
    let secondsOverMinutesToTimeout = SECS_TO_TIMEOUT - minutesToTimeout * 60;
    let MINUTES_TO_TIMEOUT = String(minutesToTimeout).padStart(2, '0');
    let SECS_OVER_MINUTES_TO_TIMEOUT = String (secondsOverMinutesToTimeout).padStart(2, '0');
    this._messageContent = eval('`' + message + '`');
  }

  /** @private */
  _maxInactiveIntervalChanged(newValue, oldValue) {
    if (newValue && this.secondsBeforeNotification && newValue > this.secondsBeforeNotification) {
      this.activated = true;
    } else {
      this.activated = false;
    }
  }

  /** @private */
  _secondsBeforeNotificationChanged(newValue, oldValue) {
    if (newValue && this.maxInactiveInterval && newValue < this.maxInactiveInterval) {
      this.activated = true;
    } else {
      this.activated = false;
    }
  }

  /** @private */
  _handleOutsideClick(e) {
    if (this.extendSessionOnOutsideClick) {
      this._pokeServer();
    } else if (this.closeNotificationOnOutsideClick) {
      this.opened = false;
    }
  }

  /** @private */
  _pokeServer(e) {
    this.$server.pokeServer().then((result) => {
      if (result === true) {
        console.log('Idle-Notification: Server poked successfully. Session extended.');
        this.opened = false;
        this._resetTimer();
      } else {
        console.error('Could not poke the server');
      }
    });
  }

  static get is() {
    return 'idle-notification';
  }
}
customElements.define(IdleNotification.is, IdleNotification);

File: src/main/java/com/vaadin/componentfactory/IdleNotification.java
/*
 * Copyright 2000-2020 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.vaadin.componentfactory;

import java.util.Objects;

import com.vaadin.flow.component.ClientCallable;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEvent;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.DomEvent;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.CssImport;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.dependency.Uses;
import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.component.internal.AllowInert;
import com.vaadin.flow.server.VaadinSession;
import com.vaadin.flow.shared.Registration;

@Tag("idle-notification")
@JsModule("./idle-notification.js")
@CssImport("./idle-notification.css")
@CssImport(value = "./vaadin-dialog-overlay.css", themeFor = "vaadin-dialog-overlay")
@Uses(Dialog.class)
public class IdleNotification extends Component {

    public static final int DEFAULT_SECONDS_BEFORE_NOTIFICATION = 60;
    public static final String DEFAULT_BEFORE_EXPIRE_MESSAGE = "Your session will expire in less than "
            + MessageFormatting.SECS_TO_TIMEOUT + " seconds. ";
    public static final String DEFAULT_AFTER_EXPIRE_MESSAGE = "Your session has expired due to inactivity.";

    public enum MessageFormatting {
        SECS_TO_TIMEOUT("${SECS_TO_TIMEOUT}"), SECS_MAX_IDLE_TIMEOUT("${SECS_MAX_IDLE_TIMEOUT}");

        public final String label;

        MessageFormatting(String label) {
            this.label = label;
        }

        @Override
        public String toString() {
            return label;
        }
    }

    private boolean closeButtonEnabled = false;
    private boolean extendSessionButtonEnabled = false;
    private boolean redirectButtonEnabled = false;
    private String message;
    private String extendSessionButtonCaption;
    private String redirectButtonCaption;
    private String redirectButtonUrl;
    private String redirectAtTimeoutUrl;
    private boolean redirectAtTimeoutEnabled = true;
    private String afterExpiredMessage;
    private Integer maxInactiveInterval;
    private Integer secondsBeforeNotification;
    private boolean extendSessionOnOutsideClick = true;
    private boolean closeNotificationOnOutsideClick = false;

    /**
     * Default constructor. Creates an idle notification with defaults.
     * <p>
     * {@link maxInactiveInterval} is obtained via
     * {@code VaadinSession.getCurrent().getSession().getMaxInactiveInterval()}
     *
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     *
     */
    public IdleNotification() throws IllegalArgumentException {
        this(DEFAULT_BEFORE_EXPIRE_MESSAGE, DEFAULT_SECONDS_BEFORE_NOTIFICATION);
    }

    /**
     * Creates an idle notification with the default message that will be
     * displayed to the user at the specified number of seconds before session
     * timeout.
     * <p>
     * {@link maxInactiveInterval} is obtained via
     * {@code VaadinSession.getCurrent().getSession().getMaxInactiveInterval()}
     *
     * @param secondsBeforeNotification
     *            the number of seconds before session times-out at which point
     *            the notification will be displayed to the user
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     */
    public IdleNotification(int secondsBeforeNotification) throws IllegalArgumentException {
        this(DEFAULT_BEFORE_EXPIRE_MESSAGE, secondsBeforeNotification,
                VaadinSession.getCurrent().getSession().getMaxInactiveInterval());
    }

    /**
     * Creates an idle notification with the specified message that will be
     * displayed to the user at the specified number of seconds before session
     * timeout.
     * <p>
     * {@link maxInactiveInterval} is obtained via
     * {@code VaadinSession.getCurrent().getSession().getMaxInactiveInterval()}
     *
     * @param message
     *            the text of the notification
     * @param secondsBeforeNotification
     *            the number of seconds before session times-out at which point
     *            the notification will be displayed to the user
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     */
    public IdleNotification(String message, int secondsBeforeNotification) throws IllegalArgumentException {
        this(message, secondsBeforeNotification, VaadinSession.getCurrent().getSession().getMaxInactiveInterval());
    }

    /**
     * Creates an idle notification with the specified message that will be
     * displayed to the user at the specified number of seconds before session
     * timeout.
     * <p>
     * {@link maxInactiveInterval} is obtained via
     * {@code VaadinSession.getCurrent().getSession().getMaxInactiveInterval()}
     *
     * @param message
     *            the text of the notification
     * @param secondsBeforeNotification
     *            the number of seconds before session times-out at which point
     *            the notification will be displayed to the user
     * @param redirectAtTimeoutUrl
     *            The URL to redirect the page to upon session timeout
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     */
    public IdleNotification(String message, int secondsBeforeNotification, String redirectAtTimeoutUrl)
            throws IllegalArgumentException {
        this(message, secondsBeforeNotification, VaadinSession.getCurrent().getSession().getMaxInactiveInterval(),
                redirectAtTimeoutUrl);
    }

    /**
     * Creates an idle notification with the specified message that will be
     * displayed to the user at the specified number of seconds before session
     * timeout. This constructor allows the specification of an arbitrary
     * {@link maxInactiveInterval}, which doesn't have to coincide with the one
     * defined for the underlying HttpSession.
     *
     * @param message
     *            the text of the notification
     * @param secondsBeforeNotification
     *            the number of seconds before session times-out at which point
     *            the notification will be displayed to the user
     * @param maxInactiveInterval
     *            the number of seconds until the session times-out
     * @param redirectAtTimeoutUrl
     *            The URL to redirect the page to upon session timeout
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     */
    public IdleNotification(String message, int secondsBeforeNotification, int maxInactiveInterval,
            String redirectAtTimeoutUrl) throws IllegalArgumentException {
        this(message, secondsBeforeNotification, maxInactiveInterval);
        setRedirectAtTimeoutUrl(redirectAtTimeoutUrl);
    }

    /**
     * Creates an idle notification with the specified message that will be
     * displayed to the user at the specified number of seconds before session
     * timeout. This constructor allows the specification of an arbitrary
     * {@link maxInactiveInterval}, which doesn't have to coincide with the one
     * defined for the underlying HttpSession.
     *
     * @param message
     *            the text of the notification
     * @param secondsBeforeNotification
     *            the number of seconds before session times-out at which point
     *            the notification will be displayed to the user
     * @param maxInactiveInterval
     *            the number of seconds until the session times-out
     * @throws IllegalArgumentException
     *             If either {@code secondsBeforeNotification} or
     *             {@code maxInactiveInterval} value is smaller than 1 seconds,
     *             or if {@code secondsBeforeNotification} value is larger than,
     *             or equal to, {@code maxInactiveInterval}.
     */
    public IdleNotification(String message, int secondsBeforeNotification, int maxInactiveInterval)
            throws IllegalArgumentException {
        if ((secondsBeforeNotification < 1) || (maxInactiveInterval < 1)) {
            throw new IllegalArgumentException("secondsBeforeNotification & maxInactiveInterval must be both > 1");
        }
        if (secondsBeforeNotification >= maxInactiveInterval) {
            throw new IllegalArgumentException("secondsBeforeNotification must be less than maxInactiveInterval");
        }
        setMessage(message);
        setSecondsBeforeNotification(secondsBeforeNotification);
        setMaxInactiveInterval(maxInactiveInterval);
        if (afterExpiredMessage == null) {
            setAfterExpiredMessage(DEFAULT_AFTER_EXPIRE_MESSAGE);
        }
    }

    /**
     * Adds a button to close the notification without poking the server
     */
    public void addCloseButton() {
        setCloseButtonEnabled(true);
    }

    /**
     * Remove the close Button
     */
    public void removeCloseButton() {
        setCloseButtonEnabled(false);
    }

    /**
     * Get the text of the notification
     *
     * @return the text of the notification
     */
    public String getMessage() {
        return message;
    }

    /**
     * Set the text of the notification with given String
     * <p>
     * NOTE: {@link MessageFormatting} can be used to format the message
     *
     * @param message
     *            the text of the Notification
     */
    public void setMessage(String message) {
        Objects.requireNonNull(message, "message should not be null");
        this.message = message;
        getElement().setProperty("beforeExpiredMessage", message);
    }

    /**
     * Add a button to the notification that allows the user to poke the server,
     * thereby extending the session
     */
    public void addExtendSessionButton(String caption) {
        Objects.requireNonNull(caption, "caption should not be null");
        setExtendSessionButtonCaption(caption);
        setExtendSessionButtonEnabled(true);
    }

    /**
     * Remove the extend-session button
     */
    public void removeExtendSessionButton() {
        setExtendSessionButtonEnabled(false);
    }

    /**
     * Add a button to the notification that allows the user to be immediately
     * redirected to a different url (e.g. a logout url)
     */
    public void addRedirectButton(String caption, String url) {
        Objects.requireNonNull(caption, "caption should not be null");
        Objects.requireNonNull(url, "url should not be null");
        setRedirectButtonCaption(caption);
        setRedirectButtonUrl(url);
        setRedirectButtonEnabled(true);
    }

    /**
     * Remove the redirect button
     */
    public void removeRedirectButton() {
        setRedirectButtonEnabled(false);
    }

    /**
     * Get the URL to redirect the page to upon session timeout
     */
    public String getRedirectAtTimeoutUrl() {
        return redirectAtTimeoutUrl;
    }

    /**
     * The URL to redirect the page to upon session timeout. Setting this value
     * to null will disable URL redirection upon timeout
     */
    public void setRedirectAtTimeoutUrl(String redirectAtTimeoutUrl) {
        this.redirectAtTimeoutUrl = redirectAtTimeoutUrl;
        getElement().setProperty("redirectAtTimeoutUrl", redirectAtTimeoutUrl);
        setRedirectAtTimeoutEnabled(redirectAtTimeoutUrl != null);
    }

    /**
     * Get the message displayed to the user after the session expires
     */
    public String getAfterExpiredMessage() {
        return afterExpiredMessage;
    }

    /**
     * Set the message displayed to the user after the session expires
     * <p>
     *
     * NOTE: this message would only be displayed if
     * {@link #redirectAtTimeoutUrl} is not set
     */
    public void setAfterExpiredMessage(String afterExpiredMessage) {
        Objects.requireNonNull(afterExpiredMessage, "afterExpiredMessage should not be null");
        this.afterExpiredMessage = afterExpiredMessage;
        getElement().setProperty("afterExpiredMessage", afterExpiredMessage);
    }

    /**
     * Get the number of seconds until the session times-out
     * <p>
     * NOTE: This getter's return value doesn't have to coincide with the one
     * defined for the underlying HttpSession.
     */
    public int getMaxInactiveInterval() {
        return maxInactiveInterval;
    }

    /**
     * Set the number of seconds until the session times-out
     * <p>
     * NOTE: This setter allows the specification of an arbitrary
     * {@link maxInactiveInterval}, which doesn't have to coincide with the one
     * defined for the underlying HttpSession.
     *
     * @throws IllegalArgumentException
     *             If {@code maxInactiveInterval} value is smaller than 1
     *             seconds, or if {@code secondsBeforeNotification} value is
     *             larger than, or equal to, {@code maxInactiveInterval}.
     */
    public void setMaxInactiveInterval(int maxInactiveInterval) throws IllegalArgumentException {
        if ((maxInactiveInterval < 1)) {
            throw new IllegalArgumentException("maxInactiveInterval must be > 1");
        }
        if ((secondsBeforeNotification != null) && (secondsBeforeNotification >= maxInactiveInterval)) {
            throw new IllegalArgumentException("secondsBeforeNotification must be less than maxInactiveInterval");
        }
        this.maxInactiveInterval = maxInactiveInterval;
        getElement().setProperty("maxInactiveInterval", maxInactiveInterval);
    }

    /**
     * Gets the number of seconds before session times-out at which point the
     * notification will be displayed to the user
     */
    public int getSecondsBeforeNotification() {
        return secondsBeforeNotification;
    }

    /**
     * Sets the number of seconds before session times-out at which point the
     * notification will be displayed to the user
     *
     * @throws IllegalArgumentException
     *             If {@code secondsBeforeNotification} value is smaller than 1
     *             seconds, or if {@code secondsBeforeNotification} value is
     *             larger than, or equal to, {@code maxInactiveInterval}.
     */
    public void setSecondsBeforeNotification(int secondsBeforeNotification) throws IllegalArgumentException {
        if ((secondsBeforeNotification < 1)) {
            throw new IllegalArgumentException("secondsBeforeNotification must be > 1");
        }
        if ((maxInactiveInterval != null) && (secondsBeforeNotification >= maxInactiveInterval)) {
            throw new IllegalArgumentException("secondsBeforeNotification must be less than maxInactiveInterval");
        }
        this.secondsBeforeNotification = secondsBeforeNotification;
        getElement().setProperty("secondsBeforeNotification", secondsBeforeNotification);
    }

    public boolean isExtendSessionOnOutsideClick() {
        return extendSessionOnOutsideClick;
    }

    /**
     * Determines whether an outside click would poke the server (thereby
     * extending the session)
     */
    public void setExtendSessionOnOutsideClick(boolean extendSessionOnOutsideClick) {
        this.extendSessionOnOutsideClick = extendSessionOnOutsideClick;
        getElement().setProperty("extendSessionOnOutsideClick", extendSessionOnOutsideClick);
    }

    public boolean isCloseNotificationOnOutsideClick() {
        return closeNotificationOnOutsideClick;
    }

    /**
     * Determines whether an outside click would close the notification without
     * poking the server
     * <p>
     * NOTE: this would only have an effect if
     * ({@link #extendSessionOnOutsideClick} == false)
     *
     */
    public void setCloseNotificationOnOutsideClick(boolean closeNotificationOnOutsideClick) {
        this.closeNotificationOnOutsideClick = closeNotificationOnOutsideClick;
        getElement().setProperty("closeNotificationOnOutsideClick", closeNotificationOnOutsideClick);
    }

    /**
     * Adds a open handler
     */
    public Registration addOpenListener(ComponentEventListener<OpenEvent> listener) {
        return addListener(OpenEvent.class, listener);
    }

    /**
     * Adds a close handler
     */
    public Registration addCloseListener(ComponentEventListener<CloseEvent> listener) {
        return addListener(CloseEvent.class, listener);
    }

    /**
     * Adds a session extend handler
     */
    public Registration addExtendSessionListener(ComponentEventListener<ExtendSessionEvent> listener) {
        return addListener(ExtendSessionEvent.class, listener);
    }

    /**
     * Adds a redirect handler
     */
    public Registration addRedirectListener(ComponentEventListener<RedirectEvent> listener) {
        return addListener(RedirectEvent.class, listener);
    }

    /**
     * Adds a timeout handler
     */
    public Registration addTimeoutListener(ComponentEventListener<TimeoutEvent> listener) {
        return addListener(TimeoutEvent.class, listener);
    }

    /**
     * Used to extend the session from the client side
     */
    @AllowInert
    @ClientCallable
    private boolean pokeServer() {
        return true;
    }

    private boolean isCloseButtonEnabled() {
        return closeButtonEnabled;
    }

    private void setCloseButtonEnabled(boolean closeButtonEnabled) {
        this.closeButtonEnabled = closeButtonEnabled;
        getElement().setProperty("closeButtonEnabled", closeButtonEnabled);
    }

    private boolean isRedirectButtonEnabled() {
        return redirectButtonEnabled;
    }

    private void setRedirectButtonEnabled(boolean redirectButtonEnabled) {
        this.redirectButtonEnabled = redirectButtonEnabled;
        getElement().setProperty("redirectButtonEnabled", redirectButtonEnabled);
    }

    private boolean isExtendSessionButtonEnabled() {
        return extendSessionButtonEnabled;
    }

    /**
     * Determines whether an extend-session button is displayed to the user
     */
    private void setExtendSessionButtonEnabled(boolean extendSessionButtonEnabled) {
        this.extendSessionButtonEnabled = extendSessionButtonEnabled;
        getElement().setProperty("extendSessionButtonEnabled", extendSessionButtonEnabled);
    }

    private String getExtendSessionButtonCaption() {
        return extendSessionButtonCaption;
    }

    /**
     * Sets the caption of the extend-session button
     */
    private void setExtendSessionButtonCaption(String extendSessionButtonCaption) {
        Objects.requireNonNull(extendSessionButtonCaption, "extendSessionButtonCaption should not be null");
        this.extendSessionButtonCaption = extendSessionButtonCaption;
        getElement().setProperty("extendSessionButtonCaption", extendSessionButtonCaption);
    }

    private String getRedirectButtonCaption() {
        return redirectButtonCaption;
    }

    private void setRedirectButtonCaption(String redirectButtonCaption) {
        Objects.requireNonNull(redirectButtonCaption, "redirectButtonCaption should not be null");
        this.redirectButtonCaption = redirectButtonCaption;
        getElement().setProperty("redirectButtonCaption", redirectButtonCaption);
    }

    private String getRedirectButtonUrl() {
        return redirectButtonUrl;
    }

    private void setRedirectButtonUrl(String redirectButtonUrl) {
        Objects.requireNonNull(redirectButtonUrl, "redirectButtonUrl should not be null");
        this.redirectButtonUrl = redirectButtonUrl;
        getElement().setProperty("redirectButtonUrl", redirectButtonUrl);
    }

    private boolean isRedirectAtTimeoutEnabled() {
        return redirectAtTimeoutEnabled;
    }

    /**
     * Determines whether the user is redirected to the
     * {@link #redirectAtTimeoutUrl} upon session timeout.
     * <p>
     *
     * NOTE: this would only have an effect if {@link #redirectAtTimeoutUrl} is
     * set
     */
    private void setRedirectAtTimeoutEnabled(boolean redirectAtTimeoutEnabled) {
        this.redirectAtTimeoutEnabled = redirectAtTimeoutEnabled;
        getElement().setProperty("redirectAtTimeoutEnabled", redirectAtTimeoutEnabled);
    }

    @DomEvent("vaadin-idle-notification-open")
    public static class OpenEvent extends ComponentEvent<IdleNotification> {
        public OpenEvent(IdleNotification source, boolean fromClient) { super(source, fromClient); }
    }

    @DomEvent("vaadin-idle-notification-close")
    public static class CloseEvent extends ComponentEvent<IdleNotification> {
        public CloseEvent(IdleNotification source, boolean fromClient) { super(source, fromClient); }
    }

    @DomEvent("vaadin-idle-notification-extend-session")
    public static class ExtendSessionEvent extends ComponentEvent<IdleNotification> {
        public ExtendSessionEvent(IdleNotification source, boolean fromClient) { super(source, fromClient); }
    }

    @DomEvent("vaadin-idle-notification-redirect")
    public static class RedirectEvent extends ComponentEvent<IdleNotification> {
        public RedirectEvent(IdleNotification source, boolean fromClient) { super(source, fromClient); }
    }

    @DomEvent("vaadin-idle-notification-timeout")
    public static class TimeoutEvent extends ComponentEvent<IdleNotification> {
        public TimeoutEvent(IdleNotification source, boolean fromClient) { super(source, fromClient); }
    }
}
